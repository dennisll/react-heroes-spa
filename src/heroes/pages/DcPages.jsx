

import React from 'react'
import { HeroList } from '../components/HeroList'

export const DcPages = () => {
  return (
    <>
      <HeroList publisher = 'DC Comics' />
    </>
  )
}

