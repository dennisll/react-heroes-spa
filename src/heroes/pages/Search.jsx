

import queryString from 'query-string'
import { useForm } from '../../hooks/useForm'
import { useLocation, useNavigate } from 'react-router-dom';
import { getHeroByName } from '../helpers';
import { HeroCard } from '../components';

export const Search = () => {

  const navigate = useNavigate();
  const location = useLocation();

  const { q = '' } = queryString.parse(location.search);
  const heroes = getHeroByName(q);

  const showSearch = (q.length === 0);
  const showError = (q.length > 0) && heroes.length === 0;

  const { searchText, onInputChange, onResetForm } = useForm({
    searchText: q
  });

  const onSearchSubmit = (e) => {
    e.preventDefault();
    //if(searchText.trim().length <= 1) return;

    //navega a la misma pagina con el criterio de busqueda indicado
    navigate(`?q=${searchText}`); 
  }

  return (
    <>
      <div className=' row mt-2'>
        <div className="col-5">
          <h4>Searching</h4>
          <hr />
          <form aria-label ='form' onSubmit={onSearchSubmit}>
            <input
              type="text"
              placeholder='Searc a hero'
              className='form-control'
              name='searchText'
              autoComplete='off'
              value={searchText}
              onChange={onInputChange}
            />

            <button
              className='btn btr-outline-primary mt-2'

            >Search</button>
          </form>
        </div>
        <div className="col-7">
          <h4>Results</h4>
          <hr />

          {/*  { (q === '')  
          ? (<div className="alert alert-primary">Search a hero</div>) 
          : (heroes.length === 0) &&
          (<div className="alert alert-danger">There is not hero with <b> {q}</b></div>)
          } 
          */}

          <div
            className="alert alert-primary animate__animated animate__fadeIn"
            style={{ display: showSearch ? '' : 'none' }}
          >Search a hero</div>

          <div
          aria-label='displayNone'
            className="alert alert-danger animate__animated animate__fadeIn"
            style={{ display: showError ? '' : 'none' }}
          >There is not hero with <b> {q}</b></div>

          {heroes.map(hero => (<HeroCard key={hero.id} {...hero} />))}
        </div>
      </div>
    </>
  )
}


