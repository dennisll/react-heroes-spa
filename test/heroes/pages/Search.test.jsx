import { expect, describe, test, jest, beforeEach } from '@jest/globals';
import { fireEvent, render, renderHook, screen } from '@testing-library/react';
import { Search } from '../../../src/heroes/pages/Search';
import { MemoryRouter } from 'react-router-dom';

const mockedUseNavigate = jest.fn();

// aca sobreescribimos solo el useNavigate de la libreria  react-router-dom
jest.mock('react-router-dom', () => (
    {
        ...jest.requireActual('react-router-dom'),
        useNavigate: () => mockedUseNavigate,
    }
));

describe('test Search Page', () => {

    beforeEach(() => jest.clearAllMocks);

    test('debe de mostrarse correctamente con valores por defecto', () => {

        const { container } = render(
            <MemoryRouter>
                <Search />
            </MemoryRouter>
        );

        expect(container).toMatchSnapshot();
    });

    test('debe de mostrar el hero buscado', () => {

        const heroSearch = 'batman';

        render(
            <MemoryRouter initialEntries={[`/search?q=${heroSearch}`]}>
                <Search />
            </MemoryRouter>
        );

        const inputValue = screen.getByRole('textbox');
        expect(inputValue.value).toBe(heroSearch);
        const img = screen.getByRole('img');
        expect(img.src).toContain("/assets/heroes/dc-batman.jpg");
        const div = screen.getByLabelText('displayNone');
        expect(div.style.display).toBe("none");
    });

    test('debe de mostrar un error al buscar el heroe inexistente', () => {

        const heroSearch = 'fff123';

        render(
            <MemoryRouter initialEntries={[`/search?q=${heroSearch}`]}>
                <Search />
            </MemoryRouter>
        );
        //screen.debug();
        const inputValue = screen.getByRole('textbox');
        expect(inputValue.value).toBe(heroSearch);
        const div = screen.getByLabelText('displayNone');
        expect(div.style.display).toBe("");
    });

    test('debe de ir a la pantalla del heroe', () => {

        const heroSearch = 'batman';

        render(
            <MemoryRouter initialEntries={[`/search?q=`]}>
                <Search />
            </MemoryRouter>
        );
        //screen.debug();


        const input = screen.getByRole('textbox');
        const form = screen.getByLabelText('form');

        fireEvent.input(input, { target: { value: heroSearch } });
        fireEvent.submit(form);
        expect(mockedUseNavigate).toHaveBeenCalledWith(`?q=${heroSearch}`); //`?q=${heroSearch}`

    });
});
