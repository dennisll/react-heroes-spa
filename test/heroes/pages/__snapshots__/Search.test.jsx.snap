// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Search Page debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <div
    class=" row mt-2"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Searc a hero"
          type="text"
          value=""
        />
        <button
          class="btn btr-outline-primary mt-2"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Search a hero
      </div>
      <div
        aria-label="displayNone"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        There is not hero with 
        <b>
           
        </b>
      </div>
    </div>
  </div>
</div>
`;
